Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs

```diff
@@ -853,6 +854,98 @@ protected internal uint stateOffsetRelativeToDeviceRoot
+        protected virtual FourCC CalculateOptimizedControlDataType()
+        {
+            return InputStateBlock.kFormatInvalid;
+        }
+
+        /// <summary>
+        /// Apply built-in parameters changes (e.g. <see cref="AxisControl.invert"/>, others) and recompute <see cref="InputControl.optimizedControlDataType"/> for impacted controls.
+        /// </summary>
+        /// <remarks>
+        /// </remarks>
+        public void ApplyParameterChanges()
+        {
+            // First we go through all children of our own hierarchy
+            SetOptimizedControlDataTypeRecursively();
+
+            // Then we go through all parents up to the root, because our own change might influence their optimization status
+            // e.g. let's say we have a tree where root is Vector3 and children are three AxisControl
+            // And user is calling this method on AxisControl which goes from Float to NotOptimized.
+            // Then we need to also transition Vector3 to NotOptimized as well.
+
+            var currentParent = parent;
+            while (currentParent != null)
+            {
+                currentParent.SetOptimizedControlDataType();
+                currentParent = currentParent.parent;
+            }
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void SetOptimizedControlDataType()
+        {
+            // setting check need to be inline so we clear optimizations if setting is disabled after the fact
+            m_OptimizedControlDataType = InputSettings.optimizedControlsFeatureEnabled
+                ? CalculateOptimizedControlDataType()
+                : (FourCC)InputStateBlock.kFormatInvalid;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal void SetOptimizedControlDataTypeRecursively()
+        {
+            // Need to go depth-first because CalculateOptimizedControlDataType might depend on computed values of children
+            if (m_ChildCount > 0)
+                foreach (var inputControl in children)
+                    inputControl.SetOptimizedControlDataTypeRecursively();
+
+            SetOptimizedControlDataType();
+        }
+
+        // This function exists to warn users to start using ApplyParameterChanges for edge cases that were previously not intentionally supported,
+        // where control properties suddenly change underneath us without us anticipating that.
+        // This is mainly to AxisControl fields being public and capable of changing at any time even if we were not anticipated such a usage pattern.
+        // Also it's not clear if InputControl.stateBlock.format can potentially change at any time, likely not.
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        // Only do this check in debug in hope that it will be sufficient to catch any misuse during development.
+        [Conditional("DEBUG")]
+        internal void EnsureOptimizationTypeHasNotChanged()
```

Comment:
Should this function go a bit further and ensure that `optimizeddControlDataType` is correct before continuing?

Reply:
Now it sets correct value `optimizedControlDataType` after printing the error.
Fixed